lib/src/storage_impl.dart

```diff
@@ -579,15 +576,23 @@ class _MediaUploadStreamSink implements StreamSink<List<int>> {
   void _onData(List<int> data) {
     assert(_state != _stateLengthKnown);
     if (_state == _stateProbingLength) {
-      buffer.add(data);
-      _bufferLength += data.length;
-      if (_bufferLength > _maxNormalUploadLength) {
+      _buffer.add(data);
+      if (_buffer.length > _maxNormalUploadLength) {
         // Start resumable upload.
         // TODO: Avoid using another stream-controller.
         _resumableController = StreamController<List<int>>(sync: true);
-        buffer.forEach(_resumableController.add);
+        _resumableController.add(_buffer.takeBytes());
         _startResumableUpload(_resumableController.stream, _length);
         _state = _stateDecidedResumable;
+
+        // At this point, we're forwarding events to the synchronous controller,
+        // so let's also forward pause and resume requests.
+        _resumableController
+          ..onPause = _subscription.pause
+          ..onResume = _subscription.resume;
+        // We don't have to handle `onCancel`: The upload will only cancel the
+        // stream in case of errors, which we already handle by closing the
+        // subscription.
```

Comment:
This means we'll have back-pressure, right?

If uploading to GSC is slowed down because of limited bandwidth then the stream will be paused. Should we perhaps make a test case for this, or just hope it works?

Reading [resumable_media_uploader.dart](https://github.com/google/googleapis.dart/blob/master/discoveryapis_commons/lib/src/resumable_media_uploader.dart), it seems the logic is a bit non-trivial, and we would have to implement a few end-point I'm guessing, if we want to test it against localhost.

Reply:
Should we land this and follow-up?

Reply:
> Should we perhaps make a test case for this, or just hope it works?

I've tried it out in a practical scenario. I essentially ran `dd if=somefile | pv | dart run upload_tool.dart`, where `upload_tool.dart` is a small script that pipes `stdin` into a GCS bucket. With the old version I see dart consuming the entire file very quickly, with the new version I see `pv` reporting a steady stream roughly matching my internet speed.

This anecdote doesn't make a test of course :D I wasn't sure how to best test this. Write a custom http client with an artificially slow bandwidth and make sure pauses propagate back into the source stream piped into GCS? I'm happy to contribute a test for this if you could give me some pointers (regardless of whether we do this here or in a follow-up).